File: python/simple-cipher/simple_cipher.py
import random
import string

class Cipher:
    def __init__(self, key=None):
        """
        Initialize a Cipher object.

        Args:
            key (str): The encryption key. If None, a random key of at least 100 lowercase characters will be generated.
        """
        if key:
            if not key.isalpha() or not key.islower():
                raise ValueError("Invalid key. Key must contain only lowercase letters.")
            self.key = key
        else:
            self.key = ''.join(random.choice(string.ascii_lowercase) for _ in range(100))

    def encode(self, text):
        """
        Encode a message using the Caesar Cipher.

        Args:
            text (str): The message to be encoded.

        Returns:
            str: The encoded message.
        """
        result = ''
        for i, char in enumerate(text):
            shift = ord(self.key[i % len(self.key)]) - ord('a')
            encoded_char = chr(((ord(char) - ord('a') + shift) % 26) + ord('a'))
            result += encoded_char
        return result

    def decode(self, text):
        """
        Decode a message encoded with the Caesar Cipher.

        Args:
            text (str): The encoded message to be decoded.

        Returns:
            str: The decoded message.
        """
        result = ''
        for i, char in enumerate(text):
            shift = ord(self.key[i % len(self.key)]) - ord('a')
            decoded_char = chr(((ord(char) - ord('a') - shift) % 26) + ord('a'))
            result += decoded_char
        return result

File: javascript/simple-cipher/simple-cipher.js
const ALPHABET = "abcdefghijklmnopqrstuvwxyz";
const ALPHABET_LENGTH = ALPHABET.length;

export class Cipher {
  constructor(key = null) {
    this._key = key || this.generateKey(); // Initialize the key or generate a random one
  }

  // Encode plainText using the key
  encode(plainText) {
    return Array.from(plainText).map((value, index) => {
      const n =
        this.letterIndex(value) +
        this.letterIndex(this._key[index % this._key.length]); // Calculate the new index
      return ALPHABET.charAt(n % ALPHABET_LENGTH); // Get the new character
    }).join(""); // Join the encoded characters into a string
  }

  // Decode cipherText using the key
  decode(cipherText) {
    return Array.from(cipherText).map((value, index) => {
      const n = 
        this.letterIndex(value) -
        this.letterIndex(this.key[index % this._key.length]); // Calculate the new index
      return ALPHABET.charAt((n + ALPHABET_LENGTH) % ALPHABET_LENGTH); // Get the new character
    }).join(""); // Join the decoded characters into a string
  }

  get key() {
    return this._key; // Get the key
  }

  // Generate a random key of a specified length
  generateKey(length = 10) {
    return [...Array(length)].map((_) => this.randomLetter()).join("");
  }

  // Generate a random lowercase letter
  randomLetter() {
    return ALPHABET.charAt(Math.floor(Math.random() * ALPHABET_LENGTH));
  }

  // Get the index of a letter in the alphabet
  letterIndex(letter) {
    return ALPHABET.indexOf(letter);
  }
}
